dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/datasetutility/ReplaceFileHandlerIT.java

```diff
@@ -53,6 +101,138 @@ public void shouldCreateDataFile() {
+        em.flush();
+
+        DataFile initialFile = createTestDataFile(dataset.getLatestVersion(), "banner", "image/png", true);
+        em.persist(initialFile);
+
+
+        byte[] bytes = IOUtils.resourceToByteArray("images/coffeeshop.png", getClass().getClassLoader());
+        File newfile = new File(FileUtil.getFilesTempDirectory() +"/coffeeshop.png");
+        try {
+            Files.write(newfile.toPath(), bytes);
+        } catch (Exception ex) {
+            ex.printStackTrace();
+        }
+
+        DataFile newDataFile = createTestDataFile(dataset.getLatestVersion(), "coffeeshop", "image/png", false);
+        newDataFile.setStorageIdentifier("coffeeshop.png");
+        em.persist(newDataFile);
+
+        //when
+        replaceFileHandler.replaceFile(dataset.getFiles().get(0), dataset, newDataFile);
+
+        //then
+        Assert.assertEquals(2 ,dataset.getFiles().size());
+        Assert.assertEquals(1, dataset.getLatestVersion().getFileMetadatas().size());
+        Assert.assertEquals(datafile.findFileMetadataByDatasetVersionId(dataset.getLatestVersion().getId(), 2,"","").size(),
+                dataset.getLatestVersion().getFileMetadatas().size());
+        Assert.assertEquals(datafile.findFileMetadataByDatasetVersionId(dataset.getLatestVersion().getId(), 2,"","").get(0),
+                dataset.getLatestVersion().getFileMetadatas().get(0));
+    }
+    private DataFile createTestDataFile(DatasetVersion datasetVersion, String filename, String fileContentType2, boolean addToDataset) {
+        DataFile.ChecksumType checksumType = DataFile.ChecksumType.fromString(settingsService.getValueForKey(SettingsServiceBean.Key.FileFixityChecksumAlgorithm));
+        DataFile savedFile = new DataFile(fileContentType2);
+        savedFile.setModificationTime(new Timestamp(new Date().getTime()));
+        savedFile.setPermissionModificationTime(new Timestamp(new Date().getTime()));
+        savedFile.setCreateDate(new Timestamp(new Date().getTime()));
+        FileMetadata fmd = new FileMetadata();
+        fmd.setLabel(filename);
+
+        FileTermsOfUse termsOfUse = new FileTermsOfUse();
+        termsOfUse.setLicense(licenseDAO.findFirstActive());
+        fmd.setTermsOfUse(termsOfUse);
+        termsOfUse.setFileMetadata(fmd);
+
+        fmd.setTermsOfUseForm(termsOfUseFormMapper.mapToForm(termsOfUse));
+
+        if(addToDataset) {
+            savedFile.setOwner(datasetVersion.getDataset());
+        }
+
+        fmd.setDataFile(savedFile);
+        savedFile.getFileMetadatas().add(fmd);
+
+        if(addToDataset) {
+            if (datasetVersion.getFileMetadatas() == null) {
+                datasetVersion.setFileMetadatas(new ArrayList<>());
+            }
+            datasetVersion.addFileMetadata(fmd);
+            fmd.setDatasetVersion(datasetVersion);
+            datasetVersion.getDataset().getFiles().add(savedFile);
+        }
+        em.persist(fmd);
+
+        datafile.generateStorageIdentifier(savedFile);
```

Comment:
I would skip service method invocation and simply make: `savedFile.setStorageIdentifier(storageId)` - storageId given as method param